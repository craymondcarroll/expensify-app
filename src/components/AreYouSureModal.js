import React from 'react';
import Modal from 'react-modal';


const AreYouSureModal = (props) => {

    console.log("Modal Props");
    console.log(props);

    return (

      <Modal
        isOpen = {props.isOpen}
        onRequestClose={props.handleCloseModal}
        contentLabel={props.modalMessage}
        closeTimeoutMS={300}
        shouldCloseOnOverlayClick={false}
        className="modal"
      >



          <div>
              <h3 className='modal__title'>{props.modalMessage}</h3>

              <div className="modal_button_space">
                 <button className='button button_modal button_danger' onClick={props.handleCloseModal}>Yes</button>
                 <button className='button button_modal' onClick={props.handleCloseModal}>No</button>
              </div>
          </div>


      </Modal>

    );

}

export default AreYouSureModal;


/**

 import React from 'react';

 import Modal from 'react-modal';



 const OptionModal = (props) =>  (

 <Modal
 isOpen={!!props.selectedOption}
 onRequestClose={props.handleClearSelectedOption}
 contentLabel="Option Selected Modal"
 closeTimeoutMS={200}
 className='modal'
 >
 <h3 className='modal__title'>Selected Option: </h3>
 {props.selectedOption  &&  <p className='modal__body'>{props.selectedOption}</p>}
 <button className='button' onClick={props.handleClearSelectedOption}>Sounds Great</button>

 </Modal>

 );



 export default OptionModal;

 */
