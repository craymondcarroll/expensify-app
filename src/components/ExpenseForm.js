import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

//-----------------------------------
// We will add component state to keep
// track of everything until they submit
// and it gets sent to redux
//--------------------------------------


export default class ExpenseForm extends React.Component {

    constructor(props) {
      super(props);

      this.state = {
          description:      props.expense ? props.expense.description : '',
          note:             props.expense ? props.expense.note : '',
          amount:           props.expense ?  (props.expense.amount / 100).toString() : '',
          createdAt:        props.expense ? moment(props.expense.createdAt) : moment(),
          calendarFocused:  false,
          error:            undefined

      };


    }






    /*******************************************
     *
     * @param e
     *******************************************/
      onDescriptionChange = (e) => {
          const description = e.target.value;

            this.setState(() =>({
                description:description
            }));
      };


    /****************************************
     * Description: On Note Changed detected
     * @param e
     ****************************************/
    onNoteChange = (e) =>{
       const note = e.target.value;

       this.setState( () => ({
           note: note
       }));

    };


    /******************************************
     * Description: On Amount Changed detected
     *
     * @param e
     ******************************************/
    onAmountChange = (e) => {
      const amount = e.target.value;

      if (amount == '' || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState( () =>({
            amount:amount
         })) ;
       }

    };


    /****************************************
     * Description: On Date Changed
     *
     * @param createdAt
     ****************************************/
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({
                createdAt: createdAt
            }));
        }
    };


    /****************************************
     * Description: On Calendar Focus Changed
     *
     * @param focused
     ****************************************/
    onFocusChange = ({focused}) => {

        this.setState( ()=>({

            calendarFocused:focused
        }));


    };


    /****************************************
     * Description: on Form Submit
     *
     * @param eevent
     ****************************************/
    onSubmit = (e) => {

        e.preventDefault();

        let hasError = undefined;

        if(!this.state.description || !this.state.amount) {
            this.setState( ()=>({error:'Please enter in Descripton and Amount'}));
        } else {
            this.setState( ()=>({error:undefined}));
            this.props.onSubmit({
                description:this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            })
        }

    };

    render() {

        return (
          <div>
            {this.state.error && <p > {this.state.error}</p>}
              <form onSubmit={this.onSubmit}>

                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    />

                <input type="text"
                       placeholder="Amount"
                       value={this.state.amount}
                       onChange={this.onAmountChange}
                />

                <SingleDatePicker
                    date ={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange ={(day) =>false}


                />

                <textarea
                    placeholder="Add a note for expense"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>

                <button>Add Expense</button>

            </form>

          </div>

        );

    }



}