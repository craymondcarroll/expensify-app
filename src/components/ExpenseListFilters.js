import React from 'react';
import {connect} from 'react-redux'
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from "../actions/filters";
import {DateRangePicker} from 'react-dates';

class ExpenseListFilters extends React.Component {

 constructor( props) {
     super(props);

     this.state = {
        calendarFocused: null
     }
 }


    onDatesChange = ({startDate,endDate}) => {
         this.props.dispatch(setStartDate(startDate));
         this.props.dispatch(setEndDate(endDate));

         
    };

    onFocusChange = (calendarFocused) => {

     this.setState( ()=> ({
        calendarFocused:calendarFocused
     }));

   };


    render() {

        return(

            <div>
                <input type="text" value={this.props.filters.text} onChange={ (e) =>{
                    this.props.dispatch(setTextFilter(e.target.value));

                }}/>

                <select value={this.props.filters.sortBy} onChange={ (e) =>{

                    switch(e.target.value) {
                        case 'date':
                            this.props.dispatch(sortByDate());
                            break;
                        case 'amount':
                            this.props.dispatch(sortByAmount());
                            break;

                        default:
                            this.props.dispatch(sortByDate());
                    }


                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>

                </select>


                    <DateRangePicker
                        startDate       ={this.props.filters.startDate}
                        startDateId     ="FilterStartDate"
                        endDate         ={this.props.filters.endDate}
                        endDateId       ="FilterEndDate"
                        onDatesChange   ={this.onDatesChange}
                        focusedInput     ={this.state.calendarFocused}
                        onFocusChange   ={this.onFocusChange}
                        numberOfMonths  ={1}
                        isOutsideRange  ={ ()=>false}
                        showClearDates  ={true}


                    />


            </div>



        );
    }

};



/***********************************************
 * Connect state to props for component.
 *
 * @param state
 * @returns {{filters: filtersReducer|demoState.filters|{text, sortBy, startDate, endDate}|Function|*}}
 ***********************************************/
const mapStateToProps = (state) => {
    return {
       filters: state.filters
    };
};


export default connect(mapStateToProps)(ExpenseListFilters);
