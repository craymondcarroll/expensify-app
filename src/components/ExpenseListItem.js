import React from 'react';
import {Link} from 'react-router-dom';


/**
 *
 * @param props
 * @returns {*}
 * @constructor
 */
const ExpenseListItem = (props) => {
    const {id,description,amount, createdAt} = props.expense;

    return(
        <div>
            <Link to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>{amount}</p>
            <p>{createdAt}</p>


        </div>
    );

};


    export default ExpenseListItem


