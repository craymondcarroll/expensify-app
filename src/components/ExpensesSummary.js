import React from 'react';
import {connect} from 'react-redux'
import selectExpenses from "../selectors/expenses";
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral'


export const ExpensesSummary =({expenseCount, expensesTotal}) => {

    const expenseWord = expenseCount ===1 ? 'expense' : 'expenses'
    const formattedAmount = numeral(expensesTotal / 100).format('$0,0.00');

    return(<div><h1>Viewing {expenseCount} {expenseWord} totaling {formattedAmount}</h1></div>);

};



//-------------------------------------------
// Get Expenses From State so we can get total
// and see how many expenses are being displayed
//--------------------------------------------
const mapStateToProps = (state) => {

    const visExpenses =  selectExpenses(state.expenses,state.filters);
    const expTotalInfo = expensesTotal(visExpenses);

    return {
        expenseCount: expTotalInfo.expNum,
        expensesTotal: expTotalInfo.total
    };
};


export default connect(mapStateToProps)(ExpensesSummary);