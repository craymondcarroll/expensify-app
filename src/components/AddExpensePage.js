import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import {addExpense} from "../actions/expenses";


const AddExpensePage = (props) => (

    <div>
      <h1>Add Expense</h1>
        <ExpenseForm

            onSubmit={ (expense) => {
                props.dispatch(addExpense(expense));
                props.history.push('/');
            }}
        />
    </div>

);


//----------------------------------------------------------------
// The export here is a common pattern it
// is the same as
// const ConnectedExpenseList = connect((state)=>{
// Which than at bottom you export default ConnectedExpenseList
//----------------------------------------------------------------

const mapStateToProps = (state) => {
    return {
        expenses:   selectExpenses(state.expenses,state.filters)
    };
};





export default connect(mapStateToProps)(AddExpensePage);