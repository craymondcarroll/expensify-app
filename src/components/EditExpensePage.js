import React from 'react';
import {connect} from 'react-redux'
import {editExpense,removeExpense} from "../actions/expenses";
import ExpenseForm from './ExpenseForm';


const EditExpensePage = (props) => {

    return (

        <div>
            <h3>Edit Expense </h3>

            <ExpenseForm

                expense={props.expense}
                onSubmit={ (expense) => {
                  props.dispatch(editExpense(props.expense.id, expense));
                  props.history.push('/');
              }}
            />

            <button onClick={ (e) =>{

                console.log(props.expense.id);
                props.dispatch(removeExpense({id:props.expense.id}));
                props.history.push('/');

            }}>Remove</button>

       </div>

    );


}



//----------------------------------------------------------------
// The export here is a common pattern it
// We have access to the expenses and to the Expense ID passed in
// we will use the array find command to match the ID's and return
// the correct expense object
//
//----------------------------------------------------------------

const mapStateToProps = (state,props) => {
    return {
        expense: state.expenses.find( (expense) => {
            return expense.id === props.match.params.id;
        })
    };
};



export default connect(mapStateToProps)(EditExpensePage)
