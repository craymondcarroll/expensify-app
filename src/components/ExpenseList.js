import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'



//--------------------------------------
// This is our normal component and we pass
// in props. props will have access to our store
// because we pass it in to React Redux HOC and
// a new component is created.
//-----------------------------------------------

export const ExpenseList = (props) => (

    <div>
             {props.expenses.map( (expense) => <ExpenseListItem key={expense.id} expense={expense}/>)}
    </div>


);


//---------------------------------
// This calls React Redux HOC and a
// new Component is returned with the
// Redux Store Information we want to
// Manage in this component
//------------------------------------

//----------------------------------------------------------------
// The export here is a common pattern it
// is the same as
// const ConnectedExpenseList = connect((state)=>{
// Which than at bottom you export default ConnectedExpenseList
//----------------------------------------------------------------

const mapStateToProps = (state) => {
    return {
        expenses:   selectExpenses(state.expenses,state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);


//--------------------------------------------------------------------------
// We need to export the new Component that is created by React Redux HOC,
//--------------------------------------------------------------------------
// export default ConnectedExpenseList;



