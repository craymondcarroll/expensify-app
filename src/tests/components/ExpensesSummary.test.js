import React from 'react';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import {shallow} from 'enzyme'

test("Should correctly render ExpensesSummary with 1", () => {

    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={(11500)} />);
    expect(wrapper).toMatchSnapshot();


});


test("Should correctly render ExpensesSummary with muliple", () => {

    const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={(115400)} />);
    expect(wrapper).toMatchSnapshot();



});