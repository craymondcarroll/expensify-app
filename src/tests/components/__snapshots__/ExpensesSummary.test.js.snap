// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpensesSummary with 1 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $115.00
  </h1>
</div>
`;

exports[`Should correctly render ExpensesSummary with muliple 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totaling 
    $1,154.00
  </h1>
</div>
`;
